tests/phpunit/CRM/MembershipExtras/Hook/ValidateForm/UpdateSubscriptionTest.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+use CRM_MembershipExtras_Test_Fabricator_Contact as ContactFabricator;
+use CRM_MembershipExtras_Hook_ValidateForm_UpdateSubscription as UpdateSubscription;
+use CRM_MembershipExtras_Test_Fabricator_RecurringContribution as RecurringContributionFabricator;
+
+/**
+ * Class CRM_MembershipExtras_Hook_ValidateForm_UpdateSubscriptionTest
+ *
+ * @group headless
+ */
+class CRM_MembershipExtras_Hook_ValidateForm_UpdateSubscriptionTest extends BaseHeadlessTest {
+
+  /**
+   * Form object that is being validated.
+   *
+   * @var \CRM_Contribute_Form_UpdateSubscription
+   */
+  private $form = NULL;
+
+  /**
+   * Array of errors.
+   *
+   * @var array
+   */
+  private $errors;
+
+  public function setUp() {
+    $this->setupUpdateSubscriptionForm();
+    $this->errors = [];
+  }
+
+  public function testErrorIsThrownWhenNextContribDateDayIsAbove28ForMonthlyMembership() {
+    $fields = [];
+    $recurringContribution = $this->createRecurContribution();
+    $this->form->set('crid', $recurringContribution['id']);
+    $fields['next_sched_contribution_date'] = '2020-12-31';
+    $fields['cycle_day'] = $recurringContribution['cycle_day'];
+
+    $updateSubscriptionValidation = new UpdateSubscription($this->form, $fields, $this->errors);
+    $updateSubscriptionValidation->validate();
+
+    $this->assertArrayHasKey('next_sched_contribution_date', $this->errors);
+    $this->assertEquals($this->errors['next_sched_contribution_date'], UpdateSubscription::INVALID_NEXT_CONTRIB_DATE_MONTH);
+  }
+
+  public function testErrorIsNotThrownWhenNextContribDateDayIsBelow29ForMonthlyMembership() {
+    $fields = [];
+    $recurringContribution = $this->createRecurContribution();
+    $this->form->set('crid', $recurringContribution['id']);
+    $fields['next_sched_contribution_date'] = '2020-12-28';
+    $fields['cycle_day'] = $recurringContribution['cycle_day'];
+
+    $updateSubscriptionValidation = new UpdateSubscription($this->form, $fields, $this->errors);
+    $updateSubscriptionValidation->validate();
+
+    $this->assertArrayNotHasKey('next_sched_contribution_date', $this->errors);
+  }
+
+  public function testErrorIsThrownWhenNextContribDateDayIsFeb29ForAnnualMembership() {
```

Comment:
for this specific test, I think testErrorIsThrownWhenNextContribDateIsLeapDateForAnnualMembership is more clear.